### PR TITLE
[`flake8-pyi`] Implement autofix for `redundant-numeric-union` (`PYI041`)

### PR DESCRIPTION
**Pull request by sbrugman**
## Summary

This PR adds autofix for `redundant-numeric-union` (`PYI041`)

There are some comments below to explain the reasoning behind some choices that might help review.

Resolves part of https://github.com/astral-sh/ruff/issues/14185.

### CODE REVIEW

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 1]
```diff
@@ -47,16 +61,21 @@ pub struct RedundantNumericUnion {
     redundancy: Redundancy,
 }
 
-impl Violation for RedundantNumericUnion {
+impl AlwaysFixableViolation for RedundantNumericUnion {
     #[derive_message_formats]
     fn message(&self) -> String {
         let (subtype, supertype) = match self.redundancy {
+            Redundancy::IntFloatComplex => ("int | float", "complex"),
```
Comment (sbrugman) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:68]: Emitting multiple messages for the same union, such as "Use `complex` instead of `int | complex`" and  "Use `complex` instead of `float | complex`" gets confusing when there is a fix presented:

```diff
-    def bad4(self, arg: Union[float | complex, int]) -> None: ...  # PYI041
+    def bad4(self, arg: complex) -> None: ...  # PYI041
```

With `IntFloatComplex` we can assign an unique violation for each combination.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 2]
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
 
-#[derive(Debug, Clone, Copy, Eq, PartialEq)]
-enum Redundancy {
-    FloatComplex,
-    IntComplex,
-    IntFloat,
-}
-
-fn check_annotation(checker: &mut Checker, annotation: &Expr) {
-    let mut has_float = false;
```
Comment (sbrugman) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:77]: Replace `has_*` with `bitflags`, as they correspond naturally to the order of Complex > Float > Int.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 3]
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
-            "int" => has_int = true,
-            "float" => has_float = true,
-            "complex" => has_complex = true,
-            _ => {}
-        }
+        numeric_flags.seen_builtin_type(builtin_type);
     };
 
+    // Traverse the union, and remember which numeric types are found.
     traverse_union(&mut find_numeric_type, checker.semantic(), annotation);
```
Comment (sbrugman) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:100]: The second loop is only active when there is a violation, instead of collecting all expressions for any Union and filtering them later to avoid overhead in those cases.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 4]
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
+    FloatComplex,
+    IntComplex,
+    IntFloat,
+}
+
+impl Redundancy {
+    pub(super) fn from_numeric_flags(numeric_flags: NumericFlags) -> Option<Self> {
+        match numeric_flags.bits() {
+            // NumericFlags::INT | NumericFlags::FLOAT | NumericFlags::COMPLEX
+            0b0111 => Some(Self::IntFloatComplex),
```
Comment (sbrugman) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:177]: There must be a more elegant way of doing this, but I haven't found it yet and this works.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 5]
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
+}
+
+// Generate a [`Fix`] for two or more type expressions, e.g. `int | float | complex`.
+fn generate_pep604_fix(
+    checker: &Checker,
+    nodes: Vec<&Expr>,
+    annotation: &Expr,
+    applicability: Applicability,
+) -> Fix {
+    debug_assert!(nodes.len() >= 2, "At least two nodes required");
```
Comment (sbrugman) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:230]: The code above handles `nodes.len() < 2` separately and thus this function should never encounter these cases. Is `debug_assert!` ok to denote this?

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 6]
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
+    traverse_union(&mut remove_numeric_type, checker.semantic(), annotation);
+
+    // Mark [`Fix`] as unsafe when comments are in range.
+    let applicability = if checker.comment_ranges().intersects(annotation.range()) {
+        Applicability::Unsafe
+    } else {
+        Applicability::Safe
+    };
+
+    // Generate the flattened fix once.
```
Comment (sbrugman) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:140]: TODO(SB):     
```rs
if checker.settings.preview.is_enabled() {
```

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 7]
```diff
@@ -37,6 +44,13 @@ use crate::checkers::ast::Checker;
 /// def foo(x: float | str) -> None: ...
 /// ```
 ///
+/// ## Fix safety
+/// This rule's fix is marked as safe for most cases; however, the fix will
+/// nested unions type expressions into a single top-level union.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:49]: the fix will...?

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 8]
```diff
@@ -37,6 +44,13 @@ use crate::checkers::ast::Checker;
...
+///
+/// The fix is marked as unsafe when comments are present within the type
+/// expression.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:52]: What's the reason that the fix is marked as unsafe when comments are present? Does it produce invalid syntax? Are there cases it can't handle?

I think our general practice has now been to not mark a fix as unsafe just because it removes comments. Instead, we stated that the fix could remove comments.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 4 above)
Comment (MichaReiser) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:182]: I don't think there is and it's inherent to bitflags where it isn't required that all possible combinations are valid. 

I would write this as

```
        if numeric_flags == NumericFlags::INT | NumericFlags::FLOAT | NumericFlags::COMPLEX {
            Some(Self::IntFloatComplex)
        } else if numeric_flags == NumericFlags::INT | NumericFlags::FLOAT {
            Some(Self::IntComplex)
        } else if numeric_flags == NumericFlags::FLOAT | NumericFlags::COMPLEX {
            Some(Self::FloatComplex)
        } else if numeric_flags == NumericFlags::FLOAT | NumericFlags::INT {
            Some(Self::IntFloat)
        } else {
            None
        }
```

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 9]
```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
+            Applicability::Unsafe
+        } else {
+            Applicability::Safe
+        };
+
+        // Generate the flattened fix once.
+        let fix = if let &[edit_expr] = necessary_nodes.as_slice() {
+            // Generate a [`Fix`] for a single type expression, e.g. `int`.
+            Fix::applicable_edit(
+                Edit::range_replacement(checker.generator().expr(edit_expr), annotation.range()),
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:149]: Is there a specific reason for using a generator over a text-based fix? Most fixes are text based to preserve existing formatting (and possibly comments).

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 10]
```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
-            checker.diagnostics.push(Diagnostic::new(
-                RedundantNumericUnion {
-                    redundancy: Redundancy::IntComplex,
-                },
-                annotation.range(),
-            ));
+        // `int` is always dropped, since `float` or `complex` must be present.
+        // `float` is only dropped if `complex`` is present.
+        if (builtin_type == "float" && !numeric_flags.contains(NumericFlags::COMPLEX))
+            || (builtin_type != "float" && builtin_type != "int")
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:127]: Isn't this the same as?

```suggestion
            || builtin_type == "complex"
```

Nit: It could also be worth to have a `BuiltInTy` enum with `Float`, `Int`, and `Complex` to reduce some of the string matching

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 11]
```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
+        {
+            necessary_nodes.push(expr);
+        }
+    };
+
+    // Traverse the union, and remember which numeric types are found.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:133]: This comment seems a copy-paste leftover from above. We aren't traversing the union to remember which numeric types are found, but to generate the fix.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 12]
```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
...
+    let new_expr = Expr::Subscript(ExprSubscript {
+        range: TextRange::default(),
+        value: Box::new(Expr::Name(ExprName {
+            id: Name::new(binding),
+            ctx: ExprContext::Store,
+            range: TextRange::default(),
+        })),
+        slice: Box::new(if let [elt] = nodes.as_slice() {
+            (*elt).clone()
+        } else {
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:285]: Isn't this branch dead because of the assertion above?

**Review by MichaReiser — COMMENTED** [review]
Thanks. This overall looks great. 

The main open questions are

* should use text based edits to avoid reformatting unchanged code and loosing inner comments. 
* Whether it's possible to always mark the fix as safe, regardless of comments

**Comment by MichaReiser** [comment]
Oh, and did you look through the ecosystem changes? If so, could you update the test plan. It's helpful to know if you did because I then don't need to carefully review all of them and can instead just click through some of them.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 10 above)
Comment (sbrugman) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:127]: Here I think the `builtin_type` is not guaranteed to be `int`/`float`/`complex`, in which case we cannot replace this condition

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 12 above)
Comment (sbrugman) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:285]: Good catch!

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 4 above)
Comment (sbrugman) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs]: Thanks, yes that looks better.

**Comment by sbrugman** [comment]
> Oh, and did you look through the ecosystem changes? If so, could you update the test plan. It's helpful to know if you did because I then don't need to carefully review all of them and can instead just click through some of them.

The ecosystem results look good.
From [this example](https://github.com/apache/superset/blob/0e165c1a21a90098adfea8efa5006e4feb2adf11/superset/utils/core.py#L349) I'm wondering if we shouldn't include return types; it appears to skip that one now.

Typeshed shows negatives in the ecosystem results, but I expect that to not be related to the PR. Maybe @AlexWaygood can help out with that.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 9 above)
Comment (sbrugman) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:149]: I've used the generator based on other rules, it could very well be that a text-based fix is more optimal.

Comment (sbrugman) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:149]: However, redundant expressions are still dropped. If they have the comment, it will be still lost.

**Comment by MichaReiser** [comment]
> From [this example](https://github.com/apache/superset/blob/0e165c1a21a90098adfea8efa5006e4feb2adf11/superset/utils/core.py#L349) I'm wondering if we shouldn't include return types; it appears to skip that one now.

Could you expand what you mean by *it appears to skip that now*. Is it not flagging this case or do you think this case shouldn't be flagged?

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 13]
```diff
@@ -37,6 +44,13 @@ use crate::checkers::ast::Checker;
 /// def foo(x: float | str) -> None: ...
 /// ```
 ///
+/// ## Fix safety
+/// This rule's fix is marked as safe for most cases; however, the fix will
+/// flatten nested unions type expressions into a single top-level union.
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:49]: Reading the documentation, I get the impression the fix isn't safe because it flattens nested union expressions. But I understand that this doesn't make the fix unsafe, it's just an extension where the fix goes beyond what one might expect.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs [hunk 14]
```diff
@@ -48,15 +62,23 @@ pub struct RedundantNumericUnion {
...
+            Redundancy::IntFloatComplex => ("int | float", "complex"),
             Redundancy::FloatComplex => ("float", "complex"),
             Redundancy::IntComplex => ("int", "complex"),
             Redundancy::IntFloat => ("int", "float"),
         };
         format!("Use `{supertype}` instead of `{subtype} | {supertype}`")
     }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Remove duplicates".to_string())
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:80]: Nit: To align with the terminology used in the documentation

```suggestion
        Some("Remove redundant type".to_string())
```

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 10 above)
Comment (MichaReiser) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:127]: Oh, in case it is e.g. `str` or `bool` and I think there are unit tests that exercise this.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 9 above)
Comment (MichaReiser) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:149]: > I've used the generator based on other rules, it could very well be that a text-based fix is more optimal.

We generally prefer using text-based fixes if they are not too complicated because they preserve more comments and the formatting. But I don't think it's worth rewriting it now, or we can do so as a separate PR .

**Review by MichaReiser — APPROVED** [review]

**Comment by AlexWaygood** [comment]
> Typeshed shows negative in the ecosystem results, but I expect that to not be related to the PR.
> ~Maybe @AlexWaygood can help out with that.~
> It's probably because the example is Py3.13+

The ecosystem reports for typeshed have [often not been reproducible locally](https://github.com/astral-sh/ruff/issues/11305) and nobody's yet had a chance to look into why... if you can't repro the typeshed hits locally, I'd just ignore them!

**Comment by sbrugman** [comment]
> Could you expand what you mean by it appears to skip that now. Is it not flagging this case or do you think this case shouldn't be flagged?

After investigating it turned out that this rule hooked into `statement.rs` instead of `expression.rs` like all the other rules related to unions do. This caused false negatives on the return types and type aliases. We had no test for this. Fixed it, and added a test.

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 13 above)
Comment (sbrugman) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:48]: Exactly. I've reused the phrasing that Charlie used for another rule.

The nesting is a feature to allow cases such as these:
```python
ReadOnlyMode         = Literal["r", "r+"]
WriteAndTruncateMode = Literal["w", "w+", "wt", "w+t"]
WriteNoTruncateMode  = Literal["r+", "r+t"]
AppendMode           = Literal["a", "a+", "at", "a+t"]

AllModes = Literal[ReadOnlyMode, WriteAndTruncateMode,
                   WriteNoTruncateMode, AppendMode]
```

As a consequence this redundant nesting is also allowed, but can be safely fixed:

```python
AllModes = Literal[Literal["r", "r+"], Literal["w", "w+", "wt", "w+t"],
                   Literal["r+", "r+t"], Literal["a", "a+", "at", "a+t"]]
```

See https://typing.readthedocs.io/en/latest/spec/literal.html#legal-and-illegal-parameterizations

We probably should have a rule for this (for both `union` and `literal`).

File: crates/ruff_linter/src/checkers/ast/analyze/expression.rs [hunk 15]
```diff
@@ -96,6 +97,9 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
                     if checker.enabled(Rule::UnnecessaryTypeUnion) {
                         flake8_pyi::rules::unnecessary_type_union(checker, expr);
                     }
+                    if checker.enabled(Rule::RedundantNumericUnion) {
+                        flake8_pyi::rules::redundant_numeric_union(checker, expr);
+                    }
```
Comment (MichaReiser) [inline crates/ruff_linter/src/checkers/ast/analyze/expression.rs:102]: Oh nice find. I think we should move this change out into its own PR so that it gets its own entry in the changelog. I have to think about it if it warrants a breaking change or not.

Comment (sbrugman) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:102]: I'll move the last commit into it's own PR

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs (hunk 13 above)
Comment (charliermarsh) [reply crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs]: I tweaked the comment a bit on all three rules.
